[DPC-3887] Update Response and Confirmation Files to match ICD

chris-ronning-ny opened the pull request:
## 🎫 Ticket

https://jira.cms.gov/browse/DPC-3887

## 🛠 Changes

- Update logic in `parsers.go` to handle the new Response file format.
- Update logic in `main.go` to handle the new Confirmation file format.
- Update data model to reflect new input file. Split parsing fields and entity model into separate structs, for optimal interaction with FileMarshaler/Unmarshaler and separation of concerns.
- Update naming conventions to reflect the following:
1.  The file we receive from S3 is now known as the Response file.
2. The file we send to S3, formerly referred to as the Response file, is now known as the Confirmation file.
- Update tests across the board.
- Update `./synthetic_test_data` and localstack.

## ℹ️ Context for reviewers

(Background context, more in-depth details of the implementation, and anything else you'd like to call out or ask reviewers.)

## ✅ Acceptance Validation
Lots of test updates. Ran in local environment with a test Response file populated with arbitrary data. Confirmed successful database updates and correct Confirmation file, dropped into localstack S3.

## 🔒 Security Implications

- [ ] This PR adds a new software dependency or dependencies.
- [ ] This PR modifies or invalidates one or more of our security controls.
- [ ] This PR stores or transmits data that was not stored or transmitted before.
- [ ] This PR requires additional review of its security implications for other reasons.

If any security implications apply, add Jason Ashbaugh (GitHub username: StewGoin) as a reviewer and do not merge this PR without his approval.

chris-ronning-ny commented on lambda/opt-out-import/synthetic_test_data/P.NGD.DPC.RSP.D240123.T1122001.IN:
```diff
@@ -0,0 +1,9 @@
+4SJ0A00AA00N
+5SJ0A00AA00N
+6SJ0A00AA00Y
+7SJ0A00AA00N
```
These are all fake NPIs

chris-ronning-ny commented on lambda/opt-out-import/Makefile:
```diff
@@ -9,6 +9,9 @@ build:
 test:
 	IS_TESTING=true DB_USER_DPC_CONSENT="postgres" DB_PASS_DPC_CONSENT="dpc-safe" go test -v -covermode=count -coverprofile=coverage/unit_coverage.out
 	go tool cover -html=coverage/unit_coverage.out -o coverage/unit_coverage.html
+test-short:
+	IS_TESTING=true DB_USER_DPC_CONSENT="postgres" DB_PASS_DPC_CONSENT="dpc-safe" go test -short -v -covermode=count -coverprofile=coverage/unit_coverage.out
+	go tool cover -html=coverage/unit_coverage.out -o coverage/unit_coverage.html
```
This will probably not need to exist once we've gotten integration tests working correctly in GH Actions.

kyeah reviewed (commented):
Nice work! Going to run this locally in a bit to verify fully, but wanted to add my note about the S3 filepath.

kyeah commented on lambda/opt-out-import/main.go:
```diff
@@ -125,11 +125,17 @@ func importOptOutFile(bucket string, file string) (bool, error) {
 		return false, err
 	} else {
-		if err = uploadResponseFile(bucket, file+"_response", s3manager.NewUploader(sess).Upload, records, fixedwidth.Marshal); err != nil {
+		if err = uploadConfirmationFile(bucket, GenerateConfirmationFileName(), s3manager.NewUploader(sess).Upload, confirmationFile); err != nil {
```
Actually, I'm going to pull in @gsf since he had opinions elsewhere about what to configure in the terraform vs. hardcode here.

gsf replied:
Thanks for tagging me. Ideally, the lambda would get the key for the download file from the `Message.Records[0].s3.object.key` value in the SNS notification that triggers it. Then the upload key would be computed from that. Doing so would avoid potentially competing authorities for the file location, and allow us to pass in alternative values in our integration tests. The same could be done for the bucket name with the `Message.Records[0].s3.bucket.name` value in the notification. 

If retrieving values from that path is too great a level of effort at this point in the development, we can temporarily reference Parameter Store values, hardcode, or reference environment variables, in that order of preference.

gsf replied:
Also, a side note that in the sample notification messages I have the object key for the file to download is `bfdeft01/dpc/in/testing.txt`. Are you saying `bfdeft01/dpc/in/` is the download or upload prefix?

chris-ronning-ny replied:
@kyeah If we're going to be reading these values from the SNS notification, it looks like we'll need to pull in another library and do some vetting. Maybe worth pushing it into a new ticket.

kyeah replied:
@chris-ronning-ny The values are already coming into the function from the event details, i.e. the bucket and filename is parsed and provided to the handler function call.

Based on @gsf's comment, this would mean that instead of replacing the input `filename` entirely, you'd parse part of it out to compute the new prefix.

so if the filename is

```
bfdeft01/dpc/in/some-filename-from-ngd
```

then you'd parse the back end of that:

```
bfdeft01/dpc
```

and then append the out directory and your generated filename.

```
bfdeft01/dpc/out/your-generated-filename
```

This _also_ means that we always expect the filename _path_ to be in a particular `/in` subdirectory so we can chop that part off to create the new outbound path.

It gets kind of messy and I don't know if the tests are already designed to be okay with that, which is why I had a preference to avoid making that application-level configuration.

If the tests don't need to change it seems like a small thing to implement in this PR — if it does require test changes then yeah, worth moving to a new ticket. @gsf I know you also recommended hardcoding AWS_ASSUME_ROLE_ARN instead of having it in the infrastructure/environment variable layer and that's another change we'll have to do regardless.

gsf replied:
Agreed on the filename handling. Thanks for clarifying! 

It would be nice if we also received the assume role in the notification message. Since we don't, I still believe it should be hardcoded, using an account ID pulled in from Parameter Store. We may also need to add a conditional to assume a different role when accessing the test bucket in our account. 

Hope that makes sense! The rationale should become more clear as we get integration tests implemented. My goal is to be able to integration test (in our account) and e2e test (in BFD account) without needing to update infrastructure or redeploy the lambda between tests.

chris-ronning-ny commented:
@kyeah I updated the filename parsing and generation as you specified, see my last commit. I've also updated the way the test file gets uploaded into the localstack bucket, so that it has the filepath format we expect to see in production. Here's some terminal output to confirm that it's working as we expect on my local.

<img width="874" alt="Screenshot 2024-03-05 at 10 40 23 AM" src="https://github.com/CMSgov/dpc-app/assets/127991809/7a338e12-90e9-4d80-bbfe-4cec47863630">

kyeah commented on lambda/opt-out-import/utils.go:
```diff
@@ -16,7 +17,9 @@ func ConvertDt(s string) (time.Time, error) {
 	currentDateTime := time.Now().Format("D060102.T150405")
-	return fmt.Sprintf("P#EFT.ON.DPC.NGD.CONF.%s", currentDateTime)
+	filename := fmt.Sprintf("P#EFT.ON.DPC.NGD.CONF.%s", currentDateTime)
+	return fmt.Sprintf("%s/%s/out/%s", path[0], path[1], filename)
```
No major concerns but may be a little more flexible if you split directly on `/in/` (in case the folder structure changes to be a level deeper or something eventually):

```go
path := strings.Split(responseFile, "/in/")
return fmt.Sprintf("%s/out/%s", path[0], filename)
```

chris-ronning-ny replied:
Alright, updated. It outputs the same results, so I will merge. Thanks for the feedback!

kyeah reviewed (approved):
LGTM, thanks!
